ambry-network/src/main/java/com.github.ambry.network/CompositeNetworkClient.java

```diff
@@ -0,0 +1,98 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+
+public class CompositeNetworkClient implements NetworkClient {
+  /**
+   * Used to properly route requests to drop.
+   */
+  private final Map<Integer, ReplicaType> correlationIdToReplicaType = new HashMap<>();
+  private final EnumMap<ReplicaType, NetworkClient> childNetworkClients;
+
+  CompositeNetworkClient(EnumMap<ReplicaType, NetworkClient> childNetworkClients) {
+    this.childNetworkClients = childNetworkClients;
+  }
+
+  @Override
+  public List<ResponseInfo> sendAndPoll(List<RequestInfo> allRequestsToSend, Set<Integer> allRequestsToDrop,
+      int pollTimeoutMs) {
+    // the first item of the pair is requests to send, the second is correlation IDs to drop
+    Function<ReplicaType, Pair<List<RequestInfo>, Set<Integer>>> pairBuilder =
+        replicaType -> new Pair<>(new ArrayList<>(), new HashSet<>());
+    EnumMap<ReplicaType, Pair<List<RequestInfo>, Set<Integer>>> requestsToSendAndDropByType =
+        new EnumMap<>(ReplicaType.class);
+    for (RequestInfo requestInfo : allRequestsToSend) {
+      ReplicaType replicaType = requestInfo.getReplicaId().getReplicaType();
+      requestsToSendAndDropByType.computeIfAbsent(replicaType, pairBuilder).getFirst().add(requestInfo);
+      correlationIdToReplicaType.put(requestInfo.getRequest().getCorrelationId(), replicaType);
+    }
+    for (Integer correlationId : allRequestsToDrop) {
+      ReplicaType replicaType = correlationIdToReplicaType.get(correlationId);
+      if (replicaType != null) {
+        requestsToSendAndDropByType.computeIfAbsent(replicaType, pairBuilder).getSecond().add(correlationId);
+      }
+    }
+    List<ResponseInfo> responses = new ArrayList<>();
+    requestsToSendAndDropByType.forEach((replicaType, requestsToSendAndDrop) -> {
+      NetworkClient client = childNetworkClients.get(replicaType);
+      if (client == null) {
+        throw new IllegalStateException("No NetworkClient configured for replica type: " + replicaType);
+      }
+      responses.addAll(
+          client.sendAndPoll(requestsToSendAndDrop.getFirst(), requestsToSendAndDrop.getSecond(), pollTimeoutMs));
+    });
+    // clean up correlation ids for completed requests
+    responses.forEach(
+        response -> correlationIdToReplicaType.remove(response.getRequestInfo().getRequest().getCorrelationId()));
+    return responses;
+  }
+
+  @Override
+  public int warmUpConnections(List<DataNodeId> dataNodeIds, int connectionWarmUpPercentagePerDataNode,
+      long timeForWarmUp, List<ResponseInfo> responseInfoList) {
+    // currently warm-up is only supported for connections to ambry-server.
+    NetworkClient client = childNetworkClients.get(ReplicaType.DISK_BACKED);
+    if (client != null) {
+      return client.warmUpConnections(dataNodeIds, connectionWarmUpPercentagePerDataNode, timeForWarmUp,
+          responseInfoList);
+    } else {
+      return 0;
```

Comment:
This doesn't seem possible.  Can't we just call the method for all clients?  LocalNetworkClient returns 0 anyway.

Reply:
Sure, I can do that. Since there is no current way of filtering by replicatype of datanode (although maybe that makes sense to add in the future), we can send the full list to both clients.